### PR TITLE
Add unary inference for integer and boolean literals

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Just trying to familiarize myself with the general patterns, testing, etc.

Part of https://github.com/astral-sh/ruff/issues/12701.

### CODE REVIEW

**Comment by charliermarsh** [comment]
A few questions:

1. Is there a way to express type errors right now? E.g., `~None` is a type error.
2. Is there a float type?

**Comment by AlexWaygood** [comment]
> ## 1. Is there a way to express type errors right now? E.g., `~None` is a type error.

Yup! Here's how we emit a diagnostic for when you try to iterate over something that's not iterable, for example:

https://github.com/astral-sh/ruff/blob/5118166d21784e7c78e38ea42919ba50bb2a5142/crates/red_knot_python_semantic/src/types/infer.rs#L1241-L1251

(I'll let you grep for places where that method is used ;)

> ## 2. Is there a float type?

There's no type in our semantic model representing _literal_ floats, no. (At least, not yet.) To say "this is an arbitrary instance of `float`", you can do something like this:

https://github.com/astral-sh/ruff/blob/5118166d21784e7c78e38ea42919ba50bb2a5142/crates/red_knot_python_semantic/src/types/infer.rs#L1620

The [typing spec](https://typing.readthedocs.io/en/latest/spec/literal.html#legal-parameters-for-literal-at-type-check-time) (drawing its language from [PEP 586](https://peps.python.org/pep-0586/#legal-parameters-for-literal-at-type-check-time)) states:

> `Literal` may be parameterized with literal ints, byte and unicode strings, bools, Enum values and `None`.

`float` isn't included there, so `Literal[3.14]` isn't valid in a type annotation according to the spec.

Should we go beyond the spec, and track the literal value of `float` variables anyway (where possible)? _Maybe_. Notably, mypy does a similar kind of tracking internally when it comes to float values -- but I think this is mostly so that it can constant-fold things like `X: Final = 3.14 ** 2` in mypyc. We don't have any equivalent of mypyc in red-knot, so I'm not sure what the use case is in red-knot for tracking the value of `float` variables _unless_ we were to allow users to write things like `Literal[3.14]` in type annnotations -- but that really would just flat-out contradict the spec, so I'd be pretty uncomfortable with allowing users to do that.

> ## 3. How do you all test these changes during debug? Do you run the playground, or the CLI, etc.?

Not sure I have a great answer there; I'm still figuring this out myself!

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 1]
```diff
@@ -2209,8 +2209,25 @@ impl<'db> TypeInferenceBuilder<'db> {
         } = unary;
 
         match (op, self.infer_expression(operand)) {
+            (UnaryOp::UAdd, Type::IntLiteral(value)) => Type::IntLiteral(value),
             (UnaryOp::USub, Type::IntLiteral(value)) => Type::IntLiteral(-value),
+            (UnaryOp::Invert, Type::IntLiteral(value)) => Type::IntLiteral(-(value + 1)),
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:2214]: Can't we use Rust's bitwise logical `not` operator here, to mirror Python's bitwise logical `not` operator?

```suggestion
            (UnaryOp::Invert, Type::IntLiteral(value)) => Type::IntLiteral(!value),
```

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 2]
```diff
@@ -2209,8 +2209,25 @@ impl<'db> TypeInferenceBuilder<'db> {
         } = unary;
 
         match (op, self.infer_expression(operand)) {
+            (UnaryOp::UAdd, Type::IntLiteral(value)) => Type::IntLiteral(value),
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:2212]:
```suggestion
		let operand_ty = self.infer_expression(operand);

        match (op, operand_ty) {
            (UnaryOp::UAdd, Type::IntLiteral(_)) => operand_ty,
```

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 3]
```diff
@@ -2209,8 +2209,25 @@ impl<'db> TypeInferenceBuilder<'db> {
...
+                Type::IntLiteral(-value)
+            }
+            (UnaryOp::Invert, Type::BooleanLiteral(bool)) => {
+                let value = i64::from(bool);
+                Type::IntLiteral(-(value + 1))
+            }
+
             (UnaryOp::Not, ty) => ty.bool(self.db).negate().into_type(self.db),
+
             _ => Type::Unknown, // TODO other unary op types
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:2231]: nit: I'd probably just inline these

```suggestion
            (UnaryOp::UAdd, Type::BooleanLiteral(bool)) => Type::IntLiteral(i64::from(bool)),
            (UnaryOp::USub, Type::BooleanLiteral(bool)) => Type::IntLiteral(-i64::from(bool))
            (UnaryOp::Invert, Type::BooleanLiteral(bool)) => Type::IntLiteral(!i64::from(bool))
            (UnaryOp::Not, ty) => ty.bool(self.db).negate().into_type(self.db),
            _ => Type::Unknown, // TODO other unary op types
```

**Review by AlexWaygood — APPROVED** [review]
Nice!

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 4]
```diff
@@ -2209,8 +2209,25 @@ impl<'db> TypeInferenceBuilder<'db> {
...
+                Type::IntLiteral(value)
+            }
+            (UnaryOp::USub, Type::BooleanLiteral(bool)) => {
+                let value = i64::from(bool);
+                Type::IntLiteral(-value)
+            }
+            (UnaryOp::Invert, Type::BooleanLiteral(bool)) => {
+                let value = i64::from(bool);
+                Type::IntLiteral(-(value + 1))
+            }
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:2227]: I'm not sure this needs to be done now (and probably it actually _shouldn't_ be done now), but at some point we'll want to emit some kind of diagnostic here. It's currently deprecated at runtime. Whether it will remain deprecated is an [open question](https://discuss.python.org/t/bool-deprecation/62232), but it's nonetheless true that typeshed [decorates `bool.__invert__` with `@deprecated`](https://github.com/AlexWaygood/typeshed/blob/9f033bf439e064917e4b235b8a09b7530f182516/stdlib/builtins.pyi#L927-L928) in its `builtins` stub, which means that (once we support `@deprecated`) we'll emit a diagnostic when users attempt to invert a non-literal boolean value. For consistency, we'll want to make sure we emit the same diagnostic for literal `bool`s as we do for non-literal `bool`s, but I'm not sure how we want to do that. (Should we check the typeshed `bool` stub for each method when we're doing type inference on literal `bool`s, in case the `bool` method is decorated with `@deprecated` or similar? Or should we copy out the same deprecation method here, and hope that we can manually keep them in sync? Curious for @carljm's thoughts!)

TL;DR: let's add a TODO comment noting that eventually this should probably cause us to emit a diagnostic due to the deprecation!

Comment (carljm) [reply crates/red_knot_python_semantic/src/types/infer.rs]: I think we'll probably want to get the deprecation message from calling the typeshed method, once we support `@deprecated`.

**Comment by carljm** [comment]
> 3\. How do you all test these changes during debug? Do you run the playground, or the CLI, etc.?

I usually do most of my debugging via tests, not via CLI. There is no red-knot playground yet.
